internal/controller/monovertexrollout/monovertexrollout_progressive.go

```diff
@@ -39,43 +44,47 @@ func (r *MonoVertexRolloutReconciler) CreateUpgradingChildDefinition(ctx context
+func (r *MonoVertexRolloutReconciler) AssessUpgradingChild(ctx context.Context, rolloutObject progressive.ProgressiveRolloutObject, existingUpgradingChildDef *unstructured.Unstructured) (apiv1.AssessmentResult, error) {
 	// TODO: Create AnalysisRun for assessing the upgrading child if user creates an AnalysisTemplate and references it in their MonoVertexRollout
-	// The following code can serve as a template for what should work:
-	/*analysisRun := &argorolloutsv1.AnalysisRun{}
-	if err := r.client.Get(ctx, client.ObjectKey{Name: existingUpgradingChildDef.GetName(), Namespace: existingUpgradingChildDef.GetNamespace()}, analysisRun); err != nil {
-		if apierrors.IsNotFound(err) {
-			analysisRun := &argorolloutsv1.AnalysisRun{
-				ObjectMeta: metav1.ObjectMeta{
-					Name:      existingUpgradingChildDef.GetName(),
-					Namespace: existingUpgradingChildDef.GetNamespace(),
-					OwnerReferences: []metav1.OwnerReference{
-						*metav1.NewControllerRef(existingUpgradingChildDef, numaflowv1.MonoVertexGroupVersionKind),
-					},
-				},
-				Spec: argorolloutsv1.AnalysisRunSpec{
-					Metrics: []argorolloutsv1.Metric{
-						{
-							Name: "my-metric",
-							Provider: argorolloutsv1.MetricProvider{
-								Prometheus: &argorolloutsv1.PrometheusMetric{
-									Address: "http://prometheus.addon-metricset-ns.svc.cluster.local:9090",
-									Query:   "vector(1) == vector(2)",
-								},
-							},
-						},
-					},
-				},
-			}
-			if err = r.client.Create(ctx, analysisRun); err != nil {
+
+	mvtxRollout := rolloutObject.(*apiv1.MonoVertexRollout)
+	analysis := mvtxRollout.GetAnalysis()
+	if len(analysis.Templates) > 0 {
+		analysisRun := &argorolloutsv1.AnalysisRun{}
+		if err := r.client.Get(ctx, client.ObjectKey{Name: existingUpgradingChildDef.GetName(), Namespace: existingUpgradingChildDef.GetNamespace()}, analysisRun); err != nil {
+			if apierrors.IsNotFound(err) {
+				err := progressive.CreateAnalysisRun(ctx, analysis, existingUpgradingChildDef, r.client)
+				if err != nil {
+					return apiv1.AssessmentResultUnknown, err
+				}
+				// analysisRun is created for the first time
+				analysisStatus := mvtxRollout.GetAnalysisStatus()
+				if analysisStatus == nil {
+					return apiv1.AssessmentResultUnknown, errors.New("analysisStatus not set")
+				}
+				analysisStatus.AnalysisRunName = existingUpgradingChildDef.GetName()
+				timeNow := metav1.NewTime(time.Now())
+				analysisStatus.StartTime = &timeNow
+				mvtxRollout.SetAnalysisStatus(analysisStatus)
+			} else {
 				return apiv1.AssessmentResultUnknown, err
 			}
-		} else {
-			return apiv1.AssessmentResultUnknown, err
 		}
-	}*/
 
-	return progressive.AssessUpgradingPipelineType(ctx, existingUpgradingChildDef, progressive.AreVertexReplicasReady)
+		// assess analysisRun status and set endTime if completed
+		analysisStatus := mvtxRollout.GetAnalysisStatus()
+		// prevents us from continually setting the endTime each time the child is assessed after the analysis run completes
```

Comment:
seeing this comment - is there a reason not to just set it every time?

Reply:
Looking at this again.. I had previously added the `analysisStatus.EndTime == nil` guard when I was using `time.Now()` (during that assessment) as the EndTime which wasn't accurate to the `analysisRun.Status.CompletedAt` which I use now.

In this case, we actually don't need that guard and I can update this comment to be more accurate since we will always only update the EndTime when the AnalysisRun is completed anyway.

Reply:
that's good, thank you